spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageLocation.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.storage;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import org.springframework.util.Assert;
+
+/**
+ * Represents a Google Cloud Storage location provided by the user.
+ *
+ * @author Daniel Zou
+ */
+public final class GoogleStorageLocation {
+
+	private static final String GCS_URI_FORMAT = "gs://%s/%s";
+
+	private final String bucketName;
+
+	private final String blobName;
+
+	private final URI uri;
+
+	/**
+	 * Constructs a {@link GoogleStorageLocation} based on the provided Google Storage URI string.
+	 * The URI string is of the form: {@code gs://<BUCKET_NAME>/<PATH_TO_FILE>}
+	 *
+	 * @param gcsLocationUriString a Google Storage URI string to a bucket/folder/file.
+	 */
+	public GoogleStorageLocation(String gcsLocationUriString) {
+		try {
+			URI locationUri = new URI(gcsLocationUriString);
+
+			Assert.isTrue(
+					gcsLocationUriString.startsWith("gs://"),
+					"A Google Storage URI must start with gs://");
+
+			Assert.isTrue(
+					locationUri.getAuthority() != null,
+					"No bucket specified in the location: " + gcsLocationUriString);
+
+			this.bucketName = locationUri.getAuthority();
+			this.blobName = getBlobPathFromUri(locationUri);
+
+			// ensure that if it's a bucket handle, location ends with a '/'
+			if (this.blobName == null && !gcsLocationUriString.endsWith("/")) {
+				locationUri = new URI(gcsLocationUriString + "/");
+			}
+			this.uri = locationUri;
+		}
+		catch (URISyntaxException e) {
+			throw new IllegalArgumentException(
+					"Invalid location: " + gcsLocationUriString);
+		}
+	}
+
+	/**
+	 * Returns the Google Storage bucket name.
+	 */
+	public String getBucketName() {
+		return bucketName;
+	}
+
+	/**
+	 * Returns the path to the blob/folder relative from the bucket root. Returns the empty
+	 * String if the {@link GoogleStorageLocation} specifies a bucket itself.
+	 */
+	public String getBlobName() {
+		return blobName;
+	}
+
+	/**
+	 * Returns the GCS URI of the location.
+	 */
+	public URI uri() {
+		return this.uri;
+	}
+
+	/**
+	 * Returns the Google Storage URI string for the location.
+	 */
+	public String uriString() {
+		String processedBlobName = (blobName == null) ? "" : blobName;
+		return String.format(GCS_URI_FORMAT, bucketName, processedBlobName);
+	}
+
+	/**
+	 * Returns a {@link GoogleStorageLocation} to a bucket.
+	 * @param bucketName name of the GCS bucket
+	 * @return the {@link GoogleStorageLocation} to the location.
+	 */
+	public static GoogleStorageLocation forBucket(String bucketName) {
+		String uri = String.format(GCS_URI_FORMAT, bucketName, "");
+		return new GoogleStorageLocation(uri);
```

Comment:
Having a constructor that just takes a bucketName and a path would come in useful here. Otherwise we end up formatting a GS:// string, and then immediately parsing it.

Reply:
Done.